Adding ColumnResizer Control for Re-sizable columns/divider Issue

IbraheemOsama opened the pull request:
Adding Column Resizer Control which add resizing functionality to a Grid Column #201 
I created the Control and a Sample page to demonstrate the functionality, I added this page to the Samples.json file to be visible in the application.

Waiting your feedback

msftclas commented:
Hi __@IbraheemOsama__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.microsoft.com.
    </span>

TTYL, MSBOT;

msftclas commented:
__@IbraheemOsama__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, MSBOT;

dotMorten commented on Microsoft.Toolkit.Uwp.UI.Controls/ColumnResizer/ColumnResizer.Events.cs:
```diff
@@ -0,0 +1,63 @@
+        /// </summary>
+        private void Resizer_PointerPressed(object sender, Windows.UI.Xaml.Input.PointerRoutedEventArgs e)
+        {
+            if (!_isResizing)
```
Shouldn't there be a CapturePointer call here?

IbraheemOsama replied:
I totally removed the pointers events in the new Push

dotMorten commented on Microsoft.Toolkit.Uwp.UI.Controls/ColumnResizer/ColumnResizer.Events.cs:
```diff
@@ -0,0 +1,63 @@
+    /// <summary>
+    /// ColumnResizer is a UI control that add the resizing functionality to a Grid Column.
+    /// </summary>
+    public partial class ColumnResizer
```
Wouldn't it be better to match WPF and call it a GridSplitter? Also we should allow resizing rows and not just columns

IbraheemOsama replied:
Changed in the new Push

dotMorten commented on Microsoft.Toolkit.Uwp.UI.Controls/ColumnResizer/ColumnResizer.cs:
```diff
@@ -0,0 +1,71 @@
+        private const double DefaultOpacity = 1;
+
+        private readonly CoreCursor _resizeCursor = new CoreCursor(CoreCursorType.SizeWestEast, 1);
+        private readonly CoreCursor _arrowCursor = new CoreCursor(CoreCursorType.Arrow, 1);
```
Should these be static?

IbraheemOsama replied:
Changed to static in the new Push

dotMorten commented on Microsoft.Toolkit.Uwp.UI.Controls/ColumnResizer/ColumnResizer.cs:
```diff
@@ -0,0 +1,71 @@
+
+            // Parent Grid (Resizable) Events Registration
+            Resizable.PointerReleased += Resizable_PointerReleased;
+            Resizable.PointerMoved += Resizable_PointerMoved;
```
If you're using pointer captures, I don't think this is necessary. It also creates a risk of leaking since the parent would have a strong pointer back to this instance

IbraheemOsama replied:
Changed in the new Push

dotMorten commented on Microsoft.Toolkit.Uwp.UI.Controls/ColumnResizer/ColumnResizer.cs:
```diff
@@ -0,0 +1,71 @@
+            // Resizer Events Registration
+            _resizer.PointerPressed += Resizer_PointerPressed;
+            _resizer.PointerEntered += Resizer_PointerEntered;
+            _resizer.PointerExited += Resizer_PointerExited;
```
Why not just listen for the event on 'this' ? Is the Template part really needed?

IbraheemOsama replied:
Changed in the new Push

dotMorten commented on Microsoft.Toolkit.Uwp.UI.Controls/ColumnResizer/ColumnResizer.cs:
```diff
@@ -0,0 +1,71 @@
+    /// ColumnResizer is a UI control that add the resizing functionality to a Grid Column.
+    /// </summary>
+    [TemplatePart(Name = RESIZERNAME, Type = typeof(Grid))]
+    public partial class ColumnResizer : Control
```
I suggest calling this `GridSplitter` and inheriting from `Thumb` instead. This'll make it match WPF

IbraheemOsama replied:
I changed the grid to thumb, but couldn't inherit Thumb because it is sealed

dotMorten commented on Microsoft.Toolkit.Uwp.UI.Controls/ColumnResizer/ColumnResizer.xaml:
```diff
@@ -0,0 +1,23 @@
+                    <Grid  x:Name="Resizer"  
+                           Width="{TemplateBinding Width}" 
+                           Background="{TemplateBinding Background}">
+                        <TextBlock HorizontalAlignment="Center"        
```
Perhaps better to use `GripperBarVertical` and `GripperBarHorizontal` from the MDL font, rather than 3 pipes. See the list here: https://msdn.microsoft.com/windows/uwp/style/segoe-ui-symbol-font.
The symbol should change based on whether it's row or column splitter

IbraheemOsama replied:
changed in the new Push

dotMorten commented on Microsoft.Toolkit.Uwp.UI.Controls/ColumnResizer/ColumnResizer.cs:
```diff
@@ -0,0 +1,71 @@
+
+        /// <summary>
+        /// Override default OnApplyTemplate to capture children controls
+        /// </summary>
```
You can change all this to just say, rather than reducing doc with a limited summary
```
/// <inheritdoc />
```

IbraheemOsama replied:
Changed in the new Push

dotMorten commented on Microsoft.Toolkit.Uwp.UI.Controls/ColumnResizer/ColumnResizer.cs:
```diff
@@ -0,0 +1,71 @@
+        /// <summary>
+        /// Gets Column Resizer Container Grid
+        /// </summary>
+        private Grid Resizable => _resizer.FindVisualAscendant<Grid>();
```
Can't you just use `this` instead of _resizer ?
Would this.Parent always be a grid? If not, this control wouldn't work, since it's not in the grid it is supposed to resize.

IbraheemOsama replied:
Changed in the new Push

dotMorten commented on Microsoft.Toolkit.Uwp.UI.Controls/ColumnResizer/ColumnResizer.cs:
```diff
@@ -0,0 +1,71 @@
+        private readonly CoreCursor _resizeCursor = new CoreCursor(CoreCursorType.SizeWestEast, 1);
+        private readonly CoreCursor _arrowCursor = new CoreCursor(CoreCursorType.Arrow, 1);
+        private Grid _resizer;
+        private bool _isResizing;
```
If you use pointer capture, there shouldn't be a need for this flag

IbraheemOsama replied:
I removed the flag because of using DragDelta

dotMorten commented on Microsoft.Toolkit.Uwp.UI.Controls/ColumnResizer/ColumnResizer.Events.cs:
```diff
@@ -0,0 +1,63 @@
+        private void Resizer_PointerEntered(object sender, Windows.UI.Xaml.Input.PointerRoutedEventArgs e)
+        {
+            Window.Current.CoreWindow.PointerCursor = _resizeCursor;
+            _resizer.Opacity = EnteredOpacity;
```
This should be done using visual states in the template

IbraheemOsama replied:
I added the opacity in the Visual States but not the cursor because it is supported in Visual State in UWP

dotMorten commented on Microsoft.Toolkit.Uwp.UI.Controls/ColumnResizer/ColumnResizer.Events.cs:
```diff
@@ -0,0 +1,63 @@
+        /// <summary>
+        /// Parent Grid On Pointer Moved Event
+        /// </summary>
+        private void Resizable_PointerMoved(object sender, PointerRoutedEventArgs e)
```
I'm wondering if it would be easier to just use ManipulationDelta instead, since this is fired for drag.

dotMorten commented on Microsoft.Toolkit.Uwp.UI.Controls/ColumnResizer/ColumnResizer.Events.cs:
```diff
@@ -0,0 +1,63 @@
+                var point = e.GetCurrentPoint(Resizable);
+                if (point.Position.X - FixedOffset >= 0)
+                {
+                    CurrentColumn.Width = new GridLength(point.Position.X - FixedOffset);
```
Have you tried this with multple columns? Looking at the math, it seems to me the CurrentColumn.Width is set to the distance from the left side of the parent grid to the mouse cursor, but if there's more than one column to the left of this column resizer instance, that wouldn't work, since the previous columns' widths should be subtracted.

IbraheemOsama replied:
In the new version I tried multiple Columns and it works fine

dotMorten commented on Microsoft.Toolkit.Uwp.UI.Controls/ColumnResizer/ColumnResizer.cs:
```diff
@@ -0,0 +1,71 @@
+            get
+            {
+                var columnResizerColumnProperty = Grid.GetColumn(this);
+                return Resizable.ColumnDefinitions[columnResizerColumnProperty - 1];
```
This could throw IndexOutOfRange if `columnResizerColumnProperty` is 0, or if there aren't enough column definitions to match the index.

IbraheemOsama replied:
Solved in the new version

ScottIsAFool commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/ColumnResizer/ColumnResizerPage.xaml.cs:
```diff
@@ -0,0 +1,30 @@
+using System.Linq;
+using System.Runtime.InteropServices.WindowsRuntime;
+using Windows.Foundation;
+using Windows.Foundation.Collections;
```
Clear redundant using statements.

IbraheemOsama commented:
@dotMorten Waiting your feedback.

ScottIsAFool commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/ColumnResizer/ColumnResizerPage.xaml.cs:
```diff
@@ -0,0 +1,30 @@
+    {
+        public ColumnResizerPage()
+        {
+            this.InitializeComponent();
```
Remove usages of "this."

ScottIsAFool commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/ColumnResizer/ColumnResizerPage.xaml.cs:
```diff
@@ -0,0 +1,30 @@
+using Windows.UI.Xaml.Media;
+using Windows.UI.Xaml.Navigation;
+
+// The Blank Page item template is documented at http://go.microsoft.com/fwlink/?LinkId=234238
```
Remove empty template comments

ScottIsAFool commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/ColumnResizer/ColumnResizerPage.xaml.cs:
```diff
@@ -0,0 +1,30 @@
+namespace Microsoft.Toolkit.Uwp.SampleApp.SamplePages
+{
+    /// <summary>
+    /// An empty page that can be used on its own or navigated to within a Frame.
```
Add more meaningful xml comments.

IbraheemOsama commented:
@ScottIsAFool I applied your comments and waiting your feedback.

Thanks.

dotMorten commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/GridSplitter/GridSplitterPage.xaml:
```diff
@@ -0,0 +1,47 @@
+            HorizontalAlignment="Stretch" 
+              VerticalAlignment="Bottom">
+        </controls:GridSplitter>
+
```
In this sample I would have expected an extra row and column, and the splitters would be in the in-between columns (similar to WPF)

IbraheemOsama replied:
Sorry but I didn't get what you mean, do you mean in the demonstration sample I should add the splitters in sort of Rectangle shape with 3 rows and 3 columns and the rectangle in the middle with 2 Row splitters and 2 Columns Splitters ?

dotMorten replied:
I would have expected the GridSplitter to be assigned to `Grid.Column="1"`, there should be 3 columns, and the second textblock should have been on 'Grid.Column="2"'.
Take a look at how it works in WPF.

IbraheemOsama replied:
I added the ResizeBehavior property to the control to match WPF

dotMorten commented on Microsoft.Toolkit.Uwp.UI.Controls/GridSplitter/GridSplitter.Options.cs:
```diff
@@ -0,0 +1,29 @@
+        /// <summary>
+        /// Grid Splitter Row Type
+        /// </summary>
+        Rows
```
Is it possible to add an 'Auto' here? Similar to this: https://msdn.microsoft.com/en-us/library/system.windows.controls.gridresizedirection(v=vs.110).aspx (Auto should be first to make it default)

IbraheemOsama replied:
Added

dotMorten commented on Microsoft.Toolkit.Uwp.UI.Controls/GridSplitter/GridSplitter.cs:
```diff
@@ -0,0 +1,107 @@
+    /// Represents the control that redistributes space between columns or rows of a Grid control.
+    /// </summary>
+    [TemplatePart(Name = SPLITTERRNAME, Type = typeof(Thumb))]
+    public partial class GridSplitter : Control
```
Sorry about the Thumb control being sealed - I had overlooked that bit. I'm not sure it's really needed then. What's the reason for having any template parts? Shouldn't the class just be a ControlControl then?

IbraheemOsama replied:
I guess using a template will help other contributors to customise the content of the control in the future, I mean maybe we can give the user the option to add his icons or event controls inside the GridSplitter.

dotMorten commented on Microsoft.Toolkit.Uwp.UI.Controls/GridSplitter/GridSplitter.cs:
```diff
@@ -0,0 +1,107 @@
+        protected override void OnApplyTemplate()
+        {
+            base.OnApplyTemplate();
+            _splitter = GetTemplateChild(SPLITTERRNAME) as Thumb;
```
Before this line, events should be unhooked if _splitter already had a value.

IbraheemOsama replied:
Fixed

dotMorten commented on Microsoft.Toolkit.Uwp.UI.Controls/GridSplitter/GridSplitter.xaml:
```diff
@@ -0,0 +1,61 @@
+                                   VerticalAlignment="Center" 
+                                   FontSize="{TemplateBinding FontSize}" 
+                                   Foreground="{TemplateBinding Foreground}" 
+                                   Text="&#xE784;" FontFamily="Segoe MDL2 Assets"></TextBlock>
```
Shouldn't the Text change if the ResizeDirection changes?

IbraheemOsama replied:
Fixed

dotMorten commented on Microsoft.Toolkit.Uwp.UI.Controls/GridSplitter/GridSplitter.Events.cs:
```diff
@@ -0,0 +1,55 @@
+
+                if (newHeight > 0)
+                {
+                    CurrentRow.Height = new GridLength(newHeight);
```
Is this enough? If the row/column after the splitter isn't '*' sized, shouldn't that row/column also be resized, and not just the one before?

IbraheemOsama replied:
I used the functions used in WPF and it was handling the column/rows with fixed sizes and with '*' size.
The minHeight is automatically respected no need to recheck.

Thanks @dotMorten so much for your valuable comments

IbraheemOsama commented:
@dotMorten I used some methods that was used in WPF Grid Splitter with applying some changes to fit UWP

dotMorten commented:
@IbraheemOsama Havent looked at your latest changes but took your control for a quick spin, and found a couple of issues.
- When you resize one of the the columns over to the other column, you can stretch it so far that they disappear and you can never recover. Generally the column the gridsplitter is in is fixed-width, and one splitter shouldn't be able to go on top of another.

- The row resizer doesn't seem to have it's own row, so it sits next to rather than under the column resizer: 
![image](https://cloud.githubusercontent.com/assets/1378165/17880804/5281ae8c-68b4-11e6-9002-6cf5a117700f.png)
Again I would have expected 3 rows here and the gridsplitter assigned to the second one. In fact I don't understand why this would work at all, if the gridsplitter doesn't have its own row it belongs to.
Similarly the textblocx and the gridsplitter are assigned to the same column. That makes no sense to me.

dotMorten commented on Microsoft.Toolkit.Uwp.UI.Controls/GridSplitter/GridSplitter.Data.cs:
```diff
@@ -0,0 +1,106 @@
+        }
+
+        // Only store resize data if we are resizing
+        private class ResizeData
```
This looks an awful lot like the code from reference source here:
http://referencesource.microsoft.com/#PresentationFramework/src/Framework/System/Windows/Controls/GridSplitter.cs,844b7b95bff72024

I think this is a violation of the license agreement you signed with this PR and the license agreement of the reference source (but not entirely sure - will have to check with the experts).

dotMorten replied:
@deltakosh Could you check into this?

IbraheemOsama replied:
@dotMorten that's why I mentioned that I used some WPF splitter functions many times to check if it is possible to use or not, actually if it is violation of license agreement I was already written most of the logic for resizing 2 fixed columns and rows maybe I could revert to this point and continue working on my own code.
 
Thanks.

deltakosh replied:
I can check but I suggest to remove it and work on our own.

IbraheemOsama replied:
@deltakosh no need to check will work on my own :) 

Thanks David.

deltakosh replied:
My pleasure..Please let us know when you have an updated version,
In the meantime I'm checking with our lawyers to understand what we can and cannot do with reference code from WPF

IbraheemOsama replied:
Will push an updated version in the next 6 hours.

deltakosh replied:
No rush :)

deltakosh commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/GridSplitter/GridSplitterPage.xaml:
```diff
@@ -0,0 +1,48 @@
+        <controls:GridSplitter 
+            Grid.Column="1"
+                Width="11"
+                Background="Gray"
```
In order to test it can you add a property to bind on resize behavior and resize direction?

deltakosh commented on Microsoft.Toolkit.Uwp.UI.Controls/GridSplitter/GridSplitter.Events.cs:
```diff
@@ -0,0 +1,125 @@
+    {
+        private static void Splitter_DragCompleted(object sender, DragCompletedEventArgs e)
+        {
+            Window.Current.CoreWindow.PointerCursor = ArrowCursor;
```
we should save the previous one and restore it

deltakosh commented on Microsoft.Toolkit.Uwp.UI.Controls/GridSplitter/GridSplitter.cs:
```diff
@@ -0,0 +1,170 @@
+    /// <summary>
+    /// Represents the control that redistributes space between columns or rows of a Grid control.
+    /// </summary>
+    [TemplatePart(Name = SPLITTERRNAME, Type = typeof(Thumb))]
```
I think we are using PascalCasing for templatePart naming

deltakosh commented on Microsoft.Toolkit.Uwp.UI.Controls/GridSplitter/GridSplitter.cs:
```diff
@@ -0,0 +1,170 @@
+        private const string SPLITTERRNAME = "Splitter";
+        private const string ICONDISPLAYNAME = "IconDisplay";
+
+        // Symbol GripperBarVertical in Segoe MDL2 Assets
```
Perhaps this could be a property?

dotMorten replied:
I think it should be content, with a visual state flipping between vertical and horizontal content. We should make these symbols the default ones though.

deltakosh commented:
This is an IMPRESSIVE pull request. Thank you very much
Just small comments and we will be good for next step: the documentation :D

IbraheemOsama commented:
@deltakosh my pleasure and thanks for you kind words really motivating :)

dotMorten commented on Microsoft.Toolkit.Uwp.UI.Controls/GridSplitter/GridSplitter.Options.cs:
```diff
@@ -0,0 +1,102 @@
+                nameof(HorizontalIconText),
+                typeof(string),
+                typeof(GridSplitter),
+                new PropertyMetadata(GripperBarHorizontal));
```
If we can't go the Content route as mentioned, I think it would be better for this to be object, and bound to a ContentControl, so you can also choose to insert image, background gradients for a rounded look or anything else you want and not just a text string.

IbraheemOsama replied:
I added ContentPresenter so anyone can add whatever control he wants.

dotMorten commented on Microsoft.Toolkit.Uwp.UI.Controls/GridSplitter/GridSplitter.cs:
```diff
@@ -0,0 +1,170 @@
+    /// Represents the control that redistributes space between columns or rows of a Grid control.
+    /// </summary>
+    [TemplatePart(Name = SplitterName, Type = typeof(Thumb))]
+    [TemplatePart(Name = IconDisplayName, Type = typeof(Thumb))]
```
Are these parts really needed?
Looks like one is just used for content (why not have a content property, or let this be a content control, and the other for manipulation, but why not just have manipulation on this instance  instead of on a part?

IbraheemOsama replied:
Removed

dotMorten commented on Microsoft.Toolkit.Uwp.UI.Controls/GridSplitter/GridSplitter.cs:
```diff
@@ -0,0 +1,170 @@
+            _splitter.Loaded += GridSplitter_Loaded;
+            _splitter.DragStarted += Splitter_DragStarted;
+            _splitter.DragDelta += Splitter_DragDelta;
+            _splitter.DragCompleted += Splitter_DragCompleted;
```
Can't these just be on this instead? (You can use Manipulation* events instead of Drag* events)

IbraheemOsama replied:
Will work on the comment and get back to you, thanks.

IbraheemOsama replied:
Changed to manipulation events on the main control.

dotMorten commented:
Very nice work! Had some suggestions for a few alternatives that might simplify the control and increase customization.

deltakosh commented:
You are a machine :) I'll give it a new try this monday (unless someone else has some times).

This control deserves a great review  ^^

deltakosh commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/GridSplitter/GridSplitterPage.xaml:
```diff
@@ -0,0 +1,74 @@
+        <controls:GridSplitter 
+            Grid.Column="1"
+                Width="11"
+            ResizeBehavior="BasedOnAlignment"
```
can you add a parameter here (to be configured by property window)

IbraheemOsama replied:
Will add ResizeBehavior and ResizeDirection as properties. do you recommend other properties to add ?

deltakosh replied:
I think they are the main ones

deltakosh commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/GridSplitter/GridSplitterPage.xaml:
```diff
@@ -0,0 +1,74 @@
+                Background="Gray"
+                Foreground="White" 
+                FontSize="13">
+            <controls:GridSplitter.Element>
```
Is there a default value is nothing is specified?

IbraheemOsama replied:
ResizeBehavior default value is BasedOnAligment and resizeDirection Default value is Auto.
Should I remove them from the sample ?

deltakosh replied:
I talked about GridSplitter.Element
For simplicity sake, it could be could if you were not forced to specify it

IbraheemOsama replied:
Will remove the element inside one of the GridSplitter and will leave the other for sake of clarification.

deltakosh replied:
Excellent!

deltakosh commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/samples.json:
```diff
@@ -82,6 +82,14 @@
+      {
+        "Name": "GridSplitter",
+        "Type": "GridSplitterPage",
+        "About": "GridSplitter Represents the control that redistributes space between columns or rows of a Grid control.",
```
Represents (lower case)

deltakosh commented:
Some slight comments but LGTM

IbraheemOsama commented:
@deltakosh I've an issue regarding adding the properties, because the properties are enums inside the controls assembly the sample.cs couldn't add it directly to the UI so I added the following code
`// Search in Microsoft.Toolkit.Uwp.UI.Controls
            var controlsProxyType = GridSplitter.GridResizeDirection.Auto;
            assembly = controlsProxyType.GetType().GetTypeInfo().Assembly;

            foreach (var typeInfo in assembly.ExportedTypes)
            {
                if (typeInfo.Name == typeName)
                {
                    return typeInfo;
                }
            }`
in LookForTypeByName function and it returned the value successfully but when I started to bind the values to UI it works only the first time I change the value and brings a wrong value. I created a 2 way binding and I tried to change the Grid.Column and it was working fine and I don't know why the binding is not working fine between I changed the ResizeDirection from the code and it works fine and responded the way it should.

Should I commit the not working changes incase if anyone can support ?
   
Thanks.

deltakosh commented:
can you confirm that your control receives the updated value the second time you click on the dropdown?

IbraheemOsama commented:
It doesn't receive it, but when I change the property from code it works fine.

deltakosh commented:
Can you update the PR? I'll check what's wrong

IbraheemOsama commented:
I updated the PR.
You can find method OnResizeDirectionChange in GridSplitter.Options
Please also note that resize behavior has the same issue.
Thanks for your help :)

deltakosh commented:
Ok solved the issue with this converter:
`
public class EnumConverter: IValueConverter
    {
        private Type _enumType;

        public EnumConverter(Type enumType)
        {
            _enumType = enumType;
        }

        public object Convert(object value, Type targetType, object parameter, string language)
        {
            return value;
        }

        public object ConvertBack(object value, Type targetType, object parameter, string language)
        {
            return Enum.Parse(_enumType, value.ToString());
        }
    }
`

and I'm using it in the PropertyControl.xaml.cs code:
```
                        case PropertyKind.Enum:
                            var enumType = option.DefaultValue.GetType();
                            var comboBox = new ComboBox
                            {
                                ItemsSource = Enum.GetNames(enumType),
                                SelectedItem = option.DefaultValue.ToString()
                            };

                            converter = new EnumConverter(enumType);
                            controlToAdd = comboBox;
                            dependencyProperty = Selector.SelectedItemProperty;
```

Can you please add it to your PR?

IbraheemOsama commented:
Interesting Solution, I thought of using converters but I stuck in what should I do in the converter :) 
Sure will add the amazing fix to the PR.
Thanks for your help.

deltakosh commented:
LGTM!!! Congrats, I know we are picky but the result is great!
Can you go ahead with the doc? :)

IbraheemOsama commented:
@deltakosh I finished the documentation and made a merged PR 2 days ago :)

deltakosh commented:
Omg...I need to sleep:) we just need another reviewer and we are good to go! Congrats

deltakosh commented:
Sorry to ask but we have conflicts here :(

IbraheemOsama commented:
There is something wrong please don't merge now :)

IbraheemOsama commented:
@deltakosh I resolved the conflicts, class OptionMenuItem was removed and I added it again.

deltakosh commented:
Actually this file was removed for true:) we don't need it anymore

IbraheemOsama commented:
I removed it :)

deltakosh commented:
LGTM

IbraheemOsama commented:
@deltakosh @dotMorten What is the next step ? :)

deltakosh commented:
Waiting for another review and we merge:)

deltakosh commented:
I merged it...Please create issues if you find a problem

deltakosh commented:
It could be cool if by default the splitter presents the grip icon (even if the user do not specify a content)

deltakosh commented:
And we should also think about using your great control for the property grid in the sample app :)

IbraheemOsama commented:
@deltakosh  Will work on the grip icon and the property grid :) 
But I've a small question, I should work on the same forked branch and create a new PR for new features ?

deltakosh commented:
A new PR would be perfect:)
